File: order.py
# !/usr/bin/python3
# Copyright (c) 2018 Jeff Lund

# Node class of binary search tree
class BST(object):
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def insert(val, parent):
    if val == parent.val:
        return
    if val < parent.val: 
        if parent.left:
            insert(val, parent.left)
        else:
            parent.left = BST(val)
    else: 
        if parent.right:
            insert(val, parent.right)
        else:
            parent.right = BST(val)

# Prints pre order traversal of BST
def preOrder(parent):
    print(parent.val, end=' ')
    
    if parent.left != None:
        preOrder(parent.left)
    
    if parent.right != None:
        preOrder(parent.right)

# Prints in order traversal of BST
def inOrder(parent):
    if parent.left != None:
        inOrder(parent.left)
    
    print(parent.val, end=' ')
    
    if parent.right != None:
        inOrder(parent.right)

# Prints post order traversal of BST
def postOrder(parent):
    if parent.left != None:
        postOrder(parent.left)
    
    if parent.right != None:
        postOrder(parent.right)
    
    print(parent.val, end=' ')

f = open("test", "r")
root = BST(f.readline().strip()) # Creating head of tree
buf = f.readlines()

for x in buf:
    insert(x.strip(), root)

print("Pre order: ", end=' ' )
preOrder(root)
print("")
print("In order: ", end=' ')
inOrder(root)
print("")
print("Post order: ", end=' ')
postOrder(root)
print("")

File: order.c
/*********************************************************
 * Copyright (c) 2018 Jeff Lund
 * Prints out a binary search tree in pre-order, in-order
 * and post-order traversals
 *********************************************************/
#include <stdio.h>
#include <stdlib.h>

// Node struct for BST
typedef struct node {
    int val;
    struct node *left;
    struct node *right;   
} node;

// Creates and returns a new node
node* createNode(int v) {
    node *temp = malloc(sizeof(node));
    temp->left = NULL;
    temp->right = NULL;
    temp->val = v;
    return temp;
} 

// Inserts a value into a BST
void insert(node *root, int v) {
    if(v == root->val)
        return;
    else if(v < root->val) {
        if(root->left == NULL) {
            root->left = createNode(v);
            return;
        }
        else
            insert(root->left, v);
    }
    else { // v > root->val
        if(root->right == NULL) {
            root->right = createNode(v);
            return;
        }
        else
            insert(root->right, v);
    }
}

// Prints out pre order traversal of BST
void preOrder(node* root) {
    printf("%d ", root->val);
    if(root->left != NULL)
        preOrder(root->left);
    if(root->right != NULL)
        preOrder(root->right);
}

// Prints out in order traversal of BST
void inOrder(node *root) {
    if(root->left != NULL)
        inOrder(root->left);
    printf("%d ", root->val);
    if(root->right != NULL)
        inOrder(root->right);
}

// Prints out post order traversal of BST
void postOrder(node *root) {
    if(root->left != NULL)
        postOrder(root->left);
    if(root->right != NULL)
        postOrder(root->right);
    printf("%d ", root->val);
}

int main(void) {
    int value;
    FILE *fp = fopen("test", "r");
    fscanf(fp, "%d", &value);
    node *root = createNode(value);

    while(fscanf(fp, "%d", &value) != EOF) {
        insert(root, value);
    }

    printf("Pre order traversal: ");
    preOrder(root);
    printf("\nIn order traversal: ");
    inOrder(root);
    printf("\nPost order traversal: ");
    postOrder(root);
    printf("\n");

    return 0;
}
